cmd/tailscale/cli/admin.go

```diff
@@ -0,0 +1,155 @@
+		Subcommands: []*ffcli.Command{{
+			Name:       "get",
+			ShortUsage: "get",
+			ShortHelp:  "Downloads the HuJSON ACL file to stdout",
+			Exec:       checkAdminKey(runAdminACLGet),
+		}, {
+			Name:       "set",
+			ShortUsage: "set",
+			ShortHelp:  "Uploads the HuJSON ACL file from stdin",
+			Exec:       checkAdminKey(runAdminACLSet),
+		}},
+		Exec: runHelp,
+	}, {
+		Name:       "devices",
+		ShortUsage: "devices <subcommand> [command flags]",
+		ShortHelp:  "Manage devices in a tailnet",
+		Subcommands: []*ffcli.Command{{
+			Name:       "list",
+			ShortUsage: "list",
+			ShortHelp:  "List all devices in a tailnet",
+			Exec:       checkAdminKey(runAdminDevicesList),
+		}, {
+			Name:       "get",
+			ShortUsage: "get <id>",
+			ShortHelp:  "Get information about a specific device",
+			Exec:       checkAdminKey(runAdminDevicesGet),
+		}},
+		Exec: runHelp,
+	}},
+	Exec: runHelp,
+}
+
+var adminArgs struct {
+	tailnet string // which tailnet to operate upon
+}
+
+func checkAdminKey(f func(context.Context, string, []string) error) func(context.Context, []string) error {
+	return func(ctx context.Context, args []string) error {
+		// TODO(dsnet): We should have a subcommand or flag to manage keys.
+		// Use of an environment variable is a temporary hack.
+		key := os.Getenv("TAILSCALE_API_KEY")
+		if !strings.HasPrefix(key, "tskey-") {
+			return errors.New("no API key specified")
+		}
+		return f(ctx, key, args)
+	}
+}
+
+func runAdminACLGet(ctx context.Context, key string, args []string) error {
+	if len(args) > 0 {
+		return flag.ErrHelp
+	}
+	return adminCallAPI(ctx, key, http.MethodGet, "/v2/tailnet/"+adminArgs.tailnet+"/acl", nil, os.Stdout)
+}
+
+func runAdminACLSet(ctx context.Context, key string, args []string) error {
+	if len(args) > 0 {
+		return flag.ErrHelp
+	}
+	return adminCallAPI(ctx, key, http.MethodPost, "/v2/tailnet/"+adminArgs.tailnet+"/acl", os.Stdin, os.Stdout)
+}
+
+func runAdminDevicesList(ctx context.Context, key string, args []string) error {
+	if len(args) > 0 {
+		return flag.ErrHelp
+	}
+	return adminCallAPI(ctx, key, http.MethodGet, "/v2/tailnet/"+adminArgs.tailnet+"/devices", nil, os.Stdout)
+}
+
+func runAdminDevicesGet(ctx context.Context, key string, args []string) error {
+	if len(args) != 1 {
+		return flag.ErrHelp
+	}
+	return adminCallAPI(ctx, key, http.MethodGet, "/v2/device/"+args[0], nil, os.Stdout)
+}
+
+func adminCallAPI(ctx context.Context, key, method, path string, in io.Reader, out io.Writer) error {
+	req, err := http.NewRequestWithContext(ctx, method, tailscaleAPIURL+path, in)
+	req.SetBasicAuth(key, "")
+	if err != nil {
+		return fmt.Errorf("failed to create request: %w", err)
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to send HTTP request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to receive HTTP response: %w", err)
+	}
+	b, err = jsonfmt.Format(b)
```

Comment:
@bradfitz. I believe you suggested that the default output not be JSON, but some output tailored towards humans. However, then we have to define our own bespoke output format. For now, I'll just use JSON, but try to format it to be as human-friendly as JSON could get.

```
$ tailscale admin devices list
{"devices": [{
    "addresses":                 ["100.109.24.53", "fd7a:115c:a1e0:ab12:4843:cd96:626d:1835"],
    "authorized":                true,
    "blocksIncomingConnections": false,
    "clientVersion":             "1.12.2-tcc3810148-g2a75956bc",
    "created":                   "2021-06-24T21:42:18Z",
    "expires":                   "2021-12-21T21:42:18Z",
    "hostname":                  "diamond",
    "id":                        "xxx",
    "isExternal":                false,
    "keyExpiryDisabled":         true,
    "lastSeen":                  "2021-08-06T17:57:38Z",
    "machineKey":                "xxx",
    "name":                      "diamond.thebrokentoaster.gmail.com",
    "nodeKey":                   "xxx",
    "os":                        "linux",
    "updateAvailable":           true,
    "user":                      "thebrokentoaster@gmail.com"
}, {
    "addresses":                 ["100.76.79.70", "fd7a:115c:a1e0:ab12:4843:cd96:624c:4f46"],
    "authorized":                true,
    "blocksIncomingConnections": false,
    "clientVersion":             "1.12.1-t07819c160-g935db9f62",
    "created":                   "2021-06-04T18:36:43Z",
    "expires":                   "2021-12-01T18:36:43Z",
    "hostname":                  "RawrPhone",
    "id":                        "xxx",
    "isExternal":                false,
    "keyExpiryDisabled":         true,
    "lastSeen":                  "2021-08-06T06:03:44Z",
    "machineKey":                "xxx",
    "name":                      "rawrphone.thebrokentoaster.gmail.com",
    "nodeKey":                   "xxx",
    "os":                        "iOS",
    "updateAvailable":           true,
    "user":                      "thebrokentoaster@gmail.com"
}, {
    "addresses":                 ["100.70.151.64", "fd7a:115c:a1e0:ab12:4843:cd96:6246:9740"],
    "authorized":                true,
    "blocksIncomingConnections": false,
    "clientVersion":             "1.10.1-t6b6016130-g49e1dcc20",
    "created":                   "2021-06-20T04:54:10Z",
    "expires":                   "2021-12-17T04:54:10Z",
    "hostname":                  "snowwhite",
    "id":                        "xxx",
    "isExternal":                false,
    "keyExpiryDisabled":         true,
    "lastSeen":                  "2021-08-06T17:14:05Z",
    "machineKey":                "xxx",
    "name":                      "snowwhite.thebrokentoaster.gmail.com",
    "nodeKey":                   "xxx",
    "os":                        "macOS",
    "updateAvailable":           true,
    "user":                      "thebrokentoaster@gmail.com"
}, {
    "addresses":                 ["100.74.216.56", "fd7a:115c:a1e0:ab12:4843:cd96:624a:d838"],
    "authorized":                true,
    "blocksIncomingConnections": false,
    "clientVersion":             "1.12.1-1d7592eb117-g1fc94ccf979",
    "created":                   "2021-07-02T07:09:48Z",
    "expires":                   "2022-01-23T19:59:37Z",
    "hostname":                  "Chromebook_57F0",
    "id":                        "xxx",
    "isExternal":                false,
    "keyExpiryDisabled":         true,
    "lastSeen":                  "2021-08-06T17:58:00Z",
    "machineKey":                "xxx",
    "name":                      "pixelbook.thebrokentoaster.gmail.com",
    "nodeKey":                   "xxx",
    "os":                        "android",
    "updateAvailable":           true,
    "user":                      "thebrokentoaster@gmail.com"
}, {
    "addresses":                 ["100.75.14.109", "fd7a:115c:a1e0:ab12:4843:cd96:624b:e6d"],
    "authorized":                true,
    "blocksIncomingConnections": false,
    "clientVersion":             "1.12.2-tcc3810148-g2a75956bc",
    "created":                   "2021-06-04T18:09:04Z",
    "expires":                   "2022-01-12T22:12:42Z",
    "hostname":                  "Glass",
    "id":                        "xxx",
    "isExternal":                false,
    "keyExpiryDisabled":         true,
    "lastSeen":                  "2021-08-06T17:57:13Z",
    "machineKey":                "xxx",
    "name":                      "glass.thebrokentoaster.gmail.com",
    "nodeKey":                   "xxx",
    "os":                        "windows",
    "updateAvailable":           true,
    "user":                      "thebrokentoaster@gmail.com"
}, {
    "addresses":                 ["100.85.80.41", "fd7a:115c:a1e0:ab12:4843:cd96:6255:5029"],
    "authorized":                true,
    "blocksIncomingConnections": false,
    "clientVersion":             "1.12.2-tcc3810148-g2a75956bc",
    "created":                   "2021-08-02T21:03:25Z",
    "expires":                   "2022-01-29T21:03:25Z",
    "hostname":                  "prism",
    "id":                        "xxx",
    "isExternal":                false,
    "keyExpiryDisabled":         true,
    "lastSeen":                  "2021-08-06T05:56:29Z",
    "machineKey":                "xxx",
    "name":                      "prism.thebrokentoaster.gmail.com",
    "nodeKey":                   "xxx",
    "os":                        "linux",
    "updateAvailable":           true,
    "user":                      "thebrokentoaster@gmail.com"
}, {
    "addresses":                 ["100.75.215.88", "fd7a:115c:a1e0:ab12:4843:cd96:624b:d758"],
    "authorized":                true,
    "blocksIncomingConnections": false,
    "clientVersion":             "1.12.2-tcc3810148-g2a75956bc",
    "created":                   "2021-07-24T01:01:03Z",
    "expires":                   "2022-01-20T01:01:03Z",
    "hostname":                  "carbonite",
    "id":                        "xxx",
    "isExternal":                false,
    "keyExpiryDisabled":         true,
    "lastSeen":                  "2021-08-06T17:57:57Z",
    "machineKey":                "xxx",
    "name":                      "carbonite.thebrokentoaster.gmail.com",
    "nodeKey":                   "xxx",
    "os":                        "linux",
    "updateAvailable":           true,
    "user":                      "thebrokentoaster@gmail.com"
}]}
```

Reply:
To further improve human readability, we could check whether this is a TTY and colorize the output.

Reply:
I like how this JSON is formatted.